File: windows/src/AgentIsland/Trigger/TriggerSafetyStore.cs
using System.ComponentModel;
using System.IO;
using AgentIsland.Core;

namespace AgentIsland.Trigger;

/// The guardrails around unattended resume: a kill switch and a per-project
/// allow list. Both persist; blocked fires are logged, never silent.
public sealed class TriggerSafetyStore : INotifyPropertyChanged
{
    public static TriggerSafetyStore Shared { get; } = new();

    private const string EnabledKey = "AgentIsland.triggerExecutionEnabled";
    private const string AllowedRootsKey = "AgentIsland.triggerAllowedRoots";

    private bool _executionEnabled;
    private HashSet<string> _allowedRoots;

    public event PropertyChangedEventHandler? PropertyChanged;

    private TriggerSafetyStore()
    {
        _executionEnabled = Preferences.Get<bool?>(EnabledKey) ?? true;
        _allowedRoots = new HashSet<string>(
            Preferences.Get<List<string>?>(AllowedRootsKey) ?? new List<string>(),
            StringComparer.OrdinalIgnoreCase);
    }

    public bool ExecutionEnabled
    {
        get => _executionEnabled;
        set
        {
            _executionEnabled = value;
            Preferences.Set(EnabledKey, value);
            Raise(nameof(ExecutionEnabled));
        }
    }

    public IReadOnlyCollection<string> AllowedRoots => _allowedRoots;

    public bool IsAllowed(string cwd)
    {
        var root = Normalized(cwd);
        return root.Length == 0 || _allowedRoots.Contains(root);
    }

    public void SetAllowed(string cwd, bool allowed)
    {
        var root = Normalized(cwd);
        if (root.Length == 0) return;
        if (allowed) _allowedRoots.Add(root); else _allowedRoots.Remove(root);
        Preferences.Set(AllowedRootsKey, _allowedRoots.OrderBy(r => r).ToList());
        Raise(nameof(AllowedRoots));
    }

    public static string Normalized(string cwd)
    {
        if (string.IsNullOrWhiteSpace(cwd)) return "";
        try { return Path.GetFullPath(cwd).TrimEnd('\\', '/'); }
        catch { return cwd.TrimEnd('\\', '/'); }
    }

    private void Raise(string name) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

File: Sources/Trigger/TriggerSafetyStore.swift
import Foundation

@MainActor
final class TriggerSafetyStore: ObservableObject {
    static let shared = TriggerSafetyStore()

    private static let enabledKey = "AgentIsland.triggerExecutionEnabled"
    private static let allowedRootsKey = "AgentIsland.triggerAllowedRoots"

    @Published var executionEnabled: Bool {
        didSet { UserDefaults.standard.set(executionEnabled, forKey: Self.enabledKey) }
    }

    @Published private(set) var allowedRoots: Set<String> {
        didSet { UserDefaults.standard.set(Array(allowedRoots).sorted(), forKey: Self.allowedRootsKey) }
    }

    private init() {
        executionEnabled = UserDefaults.standard.object(forKey: Self.enabledKey) == nil
            ? true
            : UserDefaults.standard.bool(forKey: Self.enabledKey)
        UserDefaults.standard.removeObject(forKey: "AgentIsland.triggerDryRun")
        allowedRoots = Set(UserDefaults.standard.stringArray(forKey: Self.allowedRootsKey) ?? [])
    }

    /// The allowlist entry that authorizes a trigger. A real project path
    /// trusts the whole project (all its triggers); a trigger with no cwd —
    /// common for Claude Desktop / home-dir sessions — is trusted individually
    /// by its own id. Never empty, so the approval-skipping resume still
    /// requires an explicit allow. The auto-monitoring scanner never builds a
    /// Trigger, so it can never reach this and can't auto-authorize itself.
    private func authKey(for trigger: Trigger) -> String {
        let root = normalized(trigger.cwd)
        return root.isEmpty ? "trigger:\(trigger.id)" : root
    }

    func isAllowed(_ trigger: Trigger) -> Bool {
        allowedRoots.contains(authKey(for: trigger))
    }

    func setAllowed(_ trigger: Trigger, _ allowed: Bool) {
        let key = authKey(for: trigger)
        if allowed {
            allowedRoots.insert(key)
        } else {
            allowedRoots.remove(key)
        }
    }

    func normalized(_ cwd: String) -> String {
        guard !cwd.isEmpty else { return "" }
        return (cwd as NSString).standardizingPath
    }
}
